cpp/src/experimental/louvain.cuh

```diff
@@ -15,11 +15,343 @@
+
+    auto h_src     = hasher(d_src_[index]);
+    auto h_cluster = hasher(d_dst_cluster_[d_dst_[index]]);
+
+    /*
+     * Combine the source hash and the cluster hash into a single hash value
+     *
+     * Taken from the Boost hash_combine function
+     * https://www.boost.org/doc/libs/1_35_0/doc/html/boost/hash_combine_id241013.html
+     */
+    h_src ^= h_cluster + 0x9e3779b9 + (h_src << 6) + (h_src >> 2);
+
+    return h_src;
+  }
+
+ private:
+  data_t const *d_src_;
+  data_t const *d_dst_;
+  data_t const *d_dst_cluster_;
+};
+
+//
+// Skip edges where src[e] == dst[e]
+//
+template <typename data_t>
+class skip_edge_t {
+ public:
+  skip_edge_t(rmm::device_vector<data_t> const &src, rmm::device_vector<data_t> const &dst)
+    : d_src_{src.data().get()}, d_dst_{dst.data().get()}
+  {
+  }
+
+  skip_edge_t(data_t const *src, data_t const *dst) : d_src_{src}, d_dst_{dst} {}
+
+  template <typename idx_type>
+  __device__ auto operator()(idx_type index) const
+  {
+#ifdef DEBUG
+    if (d_src_[index] == d_dst_[index]) printf("skipping edge %d\n", (int)index);
+#endif
+    return d_src_[index] == d_dst_[index];
+  }
+
+ private:
+  data_t const *d_src_;
+  data_t const *d_dst_;
+};
+
+}  // namespace detail
+
+namespace detail {
+
+template <typename vertex_t,
+          typename edge_t,
+          typename weight_t,
+          bool transposed,
+          bool multi_gpu,
+          typename view_t,
+          std::enable_if_t<multi_gpu> * = nullptr>
+std::unique_ptr<experimental::graph_t<vertex_t, edge_t, weight_t, transposed, multi_gpu>>
+create_graph(raft::handle_t const &handle,
+             rmm::device_vector<vertex_t> const &src_v,
+             rmm::device_vector<vertex_t> const &dst_v,
+             rmm::device_vector<weight_t> const &weight_v,
+             std::size_t num_local_verts,
+             experimental::graph_properties_t graph_props,
+             view_t const &view)
+{
+  std::vector<experimental::edgelist_t<vertex_t, edge_t, weight_t>> edgelist(
+    {{src_v.data().get(),
+      dst_v.data().get(),
+      weight_v.data().get(),
+      static_cast<vertex_t>(num_local_verts)}});
+
+  return std::make_unique<experimental::graph_t<vertex_t, edge_t, weight_t, transposed, multi_gpu>>(
+    handle,
+    edgelist,
+    view.get_partition(),
+    num_local_verts,
+    src_v.size(),
+    graph_props,
+    false,
+    false);
+}
+
+template <typename vertex_t,
+          typename edge_t,
+          typename weight_t,
+          bool transposed,
+          bool multi_gpu,
+          typename view_t,
+          std::enable_if_t<!multi_gpu> * = nullptr>
+std::unique_ptr<experimental::graph_t<vertex_t, edge_t, weight_t, transposed, multi_gpu>>
+create_graph(raft::handle_t const &handle,
+             rmm::device_vector<vertex_t> const &src_v,
+             rmm::device_vector<vertex_t> const &dst_v,
+             rmm::device_vector<weight_t> const &weight_v,
+             std::size_t num_local_verts,
```

Comment:
Better remove local from variable naming to avoid confusion (even though num_local_verts == num_global_verts in SG)